File: 188. Best Time to Buy and Sell Stock IV-1.cpp
class Solution {
public:
  vector<vector<array<int, 2>>> memo;
  const int inf = INT_MIN / 2;
  int maxProfit(int k, vector<int> &prices) {
    int n = prices.size();
    memo.assign(n, vector<array<int, 2>>(k + 1, {inf, inf})); // 至多

    return dfs(prices, n - 1, k, 0);
  }

  int dfs(vector<int> &prices, int i, int j, int state) {
    if (j < 0)
      return inf;
    if (i < 0)
      return state ? inf : 0;
    int &ret = memo[i][j][state];
    if (ret != inf)
      return ret;
    if (state == 1) // hold
      ret = max(dfs(prices, i - 1, j, 1),
                dfs(prices, i - 1, j - 1, 0) - prices[i]);
    else // rest
      ret = max(dfs(prices, i - 1, j, 0), dfs(prices, i - 1, j, 1) + prices[i]);
    return ret;
  }
};

File: 188. Best Time to Buy and Sell Stock IV-1.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)

        @cache
        def dfs(i, j, hold):
            if j < 0:
                return -inf
            if i < 0:
                return -inf if hold else 0
            if hold:
                return max(dfs(i - 1, j, 1), dfs(i - 1, j - 1, 0) - prices[i])
            else:
                return max(dfs(i - 1, j, 0), dfs(i - 1, j, 1) + prices[i])

        return dfs(n - 1, k, 0)
